src/result.ts

```diff
@@ -504,3 +524,8 @@ export namespace Result {
         return val instanceof Err || val instanceof Ok;
     }
 }
+
+// Utility types
+type Head<T extends any[]> = T extends [any, ...infer R] ?
+    T extends [...infer F, ...R] ? F : never : never
+type Tail<T extends any[]> = T extends [any, ...infer R] ? R : never
```

Comment:
These could likely be removed since the arrays I'm parsing above are typed as any[]. I could just as easily type the arguments as arg0: T[0] | T, argN: ...T but using these types are technically truer to what's going on.